modules/tensor_mechanics/src/actions/TensorMechanicsAction.C

```diff
@@ -478,17 +480,19 @@ TensorMechanicsAction::actOutputGeneration()
     unsigned int index = 0;
     for (auto out : _generate_output)
     {
-      std::string type;
-      if (_material_output_order[index] == "CONSTANT")
-        type = "MooseVariableConstMonomial";
-      else
-        type = "MooseVariable";
+      const auto & order = _material_output_order[index];
+      const auto & family = _material_output_family[index];
+
+      std::string type = (order == "CONSTANT" && family == "MONOMIAL")
+                             ? "MooseVariableConstMonomial"
+                             : "MooseVariable";
+
       // Create output helper aux variables
       auto params = _factory.getValidParams(type);
-      params.set<MooseEnum>("order") = _material_output_order[index];
-      params.set<MooseEnum>("family") = _material_output_family[index];
+      params.set<MooseEnum>("order") = order;
+      params.set<MooseEnum>("family") = family;
 
-      if (_material_output_family[index] == "MONOMIAL")
+      if (family == "MONOMIAL")
```

Comment:
Should we consider only `MONOMIAL` for `AuxVariable` and `LAGRANGE` for `NonlinearVariable` as output options ? I am not sure if the `L2 project` would work for any other types of interpolation function. Maybe some but not all. Throw a `mooseError`?

Reply:
Yeah, prudent to limit this until we verify.

Reply:
Thinking a bit more about this, why wouldn't arbitrary shape functions work when outputting via a nonlinear variable?

Reply:
I might be wrong, but if you take a look at all family type 

` MooseEnum family("LAGRANGE MONOMIAL HERMITE SCALAR HIERARCHIC CLOUGH XYZ SZABAB BERNSTEIN "
                   "L2_LAGRANGE L2_HIERARCHIC NEDELEC_ONE LAGRANGE_VEC MONOMIAL_VEC",
                   "LAGRANGE");`

`HERMITE` I believe only works for structure mesh (not very sure though) and undisplaced mesh. `SCALAR` won't work neither. I am not familiar with other types, such as `CLOUGH`, `BERNSTEIN`... , so I do not know if those work not or.

Reply:
HERMITE actually gives `PC failed due to FACTOR_NUMERIC_ZEROPIVOT`. Ok, I'm restricting it to `MONOMIAL` and `LAGRANGE`